storage/src/index/storage.rs

```diff
@@ -136,61 +263,70 @@ impl<T: Translator, V> Index<T, V> {
+        match entry {
+            Entry::Occupied(mut occupied_entry) => {
+                let record_ptr = occupied_entry.get_mut();
+                UpdateValueIterator {
+                    next: Some(record_ptr),
+                    entry: Some(MapEntry::Occupied(occupied_entry)),
+                    collisions_counter: &self.collisions,
+                }
+            }
+            Entry::Vacant(vacant_entry) => UpdateValueIterator {
+                next: None,
+                entry: Some(MapEntry::Vacant(vacant_entry)),
+                collisions_counter: &self.collisions,
+            },
         }
     }
 
-    /// Remove values associated with the key that match the `prune` predicate.
-    ///
-    /// If this function is never called, the amount of memory used by old values will grow
-    /// unbounded.
-    pub fn remove(&mut self, key: &[u8], prune: impl Fn(&V) -> bool) {
+    /// Retrieve all values associated with a translated key, allowing for mutation and removal.
+    pub fn remove_iter(&mut self, key: &[u8]) -> RemoveValueIterator<T::Key, V> {
         let translated_key = self.translator.transform(key);
-        let head = match self.map.get_mut(&translated_key) {
-            Some(head) => head,
-            None => return,
-        };
-
-        // Advance the head of the linked list to the first entry that will be retained, if any.
-        loop {
-            if !prune(&head.value) {
-                break;
-            }
-            self.keys_pruned.inc();
-            match head.next.take() {
-                Some(next) => {
-                    *head = *next;
-                }
-                None => {
-                    // No retained entries, so remove the key from the map.
-                    self.map.remove(&translated_key);
-                    return;
+        let entry = self.map.entry(translated_key);
+        match entry {
+            Entry::Occupied(mut occupied_entry) => {
+                let record_ptr = occupied_entry.get_mut();
+                RemoveValueIterator {
+                    can_remove: false,
+                    prev_prev: None,
+                    prev: None,
+                    next: Some(record_ptr),
+                    entry: Some(occupied_entry),
+                    pruned_counter: &self.keys_pruned,
                 }
             }
+            Entry::Vacant(_) => RemoveValueIterator {
+                can_remove: false,
+                prev_prev: None,
+                prev: None,
+                next: None,
+                entry: None,
+                pruned_counter: &self.keys_pruned,
+            },
         }
+    }
 
-        // Prune the remainder of the list.
-        let mut cursor = head;
-        while let Some(value) = cursor.next.as_ref().map(|next| &next.value) {
+    /// Remove all values associated with a translated key that match the `prune` predicate.
+    pub fn remove(&mut self, key: &[u8], prune: impl Fn(&V) -> bool) {
```

Comment:
For future work: wonder if we should add a version of this that breaks after the first match of the prune predicate

Reply:
At least iterator provides that functionality if you want it, just a bit more verbosely.